File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* root, int max_val) {
        if (root == NULL) return 0;
        max_val = max(max_val, root->val);
        if (root->val >= max_val) {
            return 1 + dfs(root->left, max_val) + dfs(root->right, max_val);
        }
        return dfs(root->left, max_val) + dfs(root->right, max_val);
    }
    int goodNodes(TreeNode* root) {
        return dfs(root, INT_MIN);
    }
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(root, maxValue):
            if not root:
                return 0
            res = 1 if root.val >= maxValue else 0
            maxValue = max(maxValue, root.val)
            res += dfs(root.left, maxValue)
            res += dfs(root.right, maxValue)
            return res
        return dfs(root, root.val)